[feat] Periodogram and PowerSpectrum Analysis

Qixian-Wang opened the pull request:
Modules for Spectrogram Analysis, PSD computing and Power Spectrum Analysis are added. 

## Key Updates
- Module `power_density_statistics`: This module provides support for three methods for power spectral density (PSD) computation:  Welch's Method, Periodogram Method, Multitaper Method.

- Module `spectrum_analysis`: This module supports the computation and visualization of absolute band power and relative band power for user-defined frequency bands.

- Module `spectrogram_analysis`: This module enables visualization of spectrograms.

## Other updates
- Module `multitaper_spectrum_statistics`: This module provides a custom function for computing PSD using the Multitaper method.

- New test cases: Developed test cases to cover typical use cases.

## Dataflow
```mermaid
graph LR
P1(signal)
P2[ButterBandpass]

A1[SpectrumAnalysisWelch]
A2[SpectrumAnalysisPeriodogram]
A3[SpectrumAnalysisMultitaper]
        
B1[[SpectrogramAnalysis Operator]]
B2[[PowerSpectrumAnalysis]]
    
P1 --> P2
P2 --> A1
P2 --> A2
P2 --> A3
    
P1 --> B1
A1 --> B2
A2 --> B2
A3 --> B2
```

Note: `ButterBandpass` is optional based on specific requirements.

## TODO

- [x] Typehint (check with `mypy`)
- [x] Test
- [x] Code Formatting/Linting
- [x] Make proper PR statement here

skim0119 commented on PeriodogramAnalysis/PeriodogramAnalysis.py:
```diff
@@ -5,35 +5,33 @@
+@dataclass
 class PeriodogramAnalysis(OperatorMixin):
     """
        A class to perform Periodogram Analysis using Welch's method and multitaper PSD.
```
Can you put the reference for this analysis?

Qixian-Wang replied:
You mean putting the reference website in docstring? Yes, I have added it.

skim0119 replied:
The text is wrong. You didn't document the information on that page yourself.
Just say `the algorithm implementation is inspired from ...`.

Qixian-Wang replied:
I see. 
Have changed.

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -59,139 +56,147 @@ def __call__(self, signal: SignalType):
+                if show:
+                    plt.show()
+                if save_path is not None:
+                    plot_path = os.path.join(save_path, f'Chunk{chunk}:Spectrogram_Channel_{channel}.png')
```
No special character in filename (:)

Qixian-Wang replied:
Deleted it

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -59,139 +56,147 @@ def __call__(self, signal: SignalType):
+                if show:
+                    plt.show()
+                if save_path is not None:
+                    plot_path = os.path.join(save_path, f"Chunk{chunk}:Comparison_figure_channel:{channel}.png")
```
No special character in file name

Qixian-Wang replied:
Deleted it

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -59,139 +56,147 @@ def __call__(self, signal: SignalType):
+        psd_dict = defaultdict(dict)
+        spec_dict = defaultdict(dict)
+        for chunk_index, signal_piece in enumerate(signal):
+            self.chunk_idx = chunk_index
             self.num_channels = signal_piece.number_of_channels
```
It is bad practice: Don't define class variable within the member function. It become hard to trace where these variables are coming from.

Qixian-Wang replied:
Get it. I have removed them.

skim0119 commented on PeriodogramAnalysis/PeriodogramAnalysis.py:
```diff
@@ -319,23 +305,21 @@ def computing_welch_bandpower(self, signal, band, channel, window_sec=None, rela
+            nperseg = self.window_length_for_welch * frequency 
         else:
             nperseg = (2 / low) * frequency
         freqs, psd = sig.welch(local_signal[:, channel], frequency, nperseg=nperseg)
```
Just use `signal[channel]`. No need to redefine `local_signal`

Qixian-Wang replied:
Sure, have modified it.

skim0119 replied:
Can you just use `signal[channel]` instead of `signal.data[:,channel]`?

Qixian-Wang replied:
Have changed. I thought it won't work previously...

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -59,139 +56,147 @@ def __call__(self, signal: SignalType):
+                    plt.savefig(plot_path, dpi=300)
+                plt.close()
 
 def multitaper_psd(x, sfreq, fmin=0.0, fmax=np.inf, bandwidth=None, adaptive=True, low_bias=True):
```
This function can be on separate file.

Qixian-Wang replied:
I have removed it to another file named MultitaperPowerSpectrum.py

skim0119 commented on PeriodogramAnalysis/PeriodogramAnalysis.py:
```diff
@@ -5,35 +5,33 @@
 
-    tag = "Periodogram Analysis"
+    exclude_channel_list: list = field(default_factory=list)
+    band_list: list = field(default_factory=lambda: [[0.5, 4], [4, 8], [8, 12], [12, 30], [30, 100]])
```
use tuple instead

Qixian-Wang replied:
Have changed it to tuple

skim0119 replied:
You can remove `field`

Qixian-Wang replied:
Ok, removed second field.

skim0119 commented on PeriodogramAnalysis/PeriodogramAnalysis.py:
```diff
@@ -50,233 +48,221 @@ def __call__(self, signal: SignalType):
+        power_dict = defaultdict(dict)
+        for chunk_idx, signal_piece in enumerate(signal):
+            self.chunk_idx = chunk_idx
             self.num_channels = signal_piece.number_of_channels
```
No need

Qixian-Wang commented:
![Chunk0:welch_periodogram_of_channel_3](https://github.com/user-attachments/assets/693de423-58c5-48f1-a16e-2284c800e044)
![Chunk1:welch_periodogram_of_channel_3](https://github.com/user-attachments/assets/3d08923e-1bd4-43e7-9e30-40d7515b7c6d)
![Chunk0:Comparison_figure_channel:3](https://github.com/user-attachments/assets/9fadef21-9608-470d-a7ea-04ce1899d190)
![Chunk0:Spectrogram_Channel_3](https://github.com/user-attachments/assets/6190caf4-0141-46e3-925d-3d155de70281)
![Chunk1:Comparison_figure_channel:3](https://github.com/user-attachments/assets/76c40944-5dac-4dce-9320-418d9238e775)
![Chunk1:Spectrogram_Channel_3](https://github.com/user-attachments/assets/a32fe2fa-4c30-405c-a7f6-a8e72d674563)

skim0119 commented:
@Qixian-Wang 
- When you run spectral analysis, you need to remove the bias to eliminate the power at 0Hz.
- Legend, xy-labels

Qixian-Wang commented:
![Chunk0_Spectrogram_Channel_3](https://github.com/user-attachments/assets/20876df3-21c9-4ca8-9057-94067681eb1a)
![Chunk0_Welch_periodogram_of_channel_3](https://github.com/user-attachments/assets/d0754131-ea4c-4bf7-af32-cae0779372d1)
I have removed the bias when Hz = 0 in Spectrogram. Also added legends to periodogram, can you specify what other labels and legends need to be changed?

skim0119 commented:
> I have removed the bias when Hz = 0 in Spectrogram. Also added legends to periodogram, can you specify what other labels and legends need to be changed?

How did you remove the bias? I don't see it in the code, and the plot still has biases.
Also I would plot the frequency axis in log scale.

Qixian-Wang commented:
![Chunk0_Spectrogram_Channel_3](https://github.com/user-attachments/assets/05638467-c72c-4685-90bb-9f4ccd7cc8e2)
What do you think of this one? I googled Spectrogram but seldom use log space though.

Qixian-Wang commented:
the code removing bias is here: https://github.com/Qixian-Wang/NEF/blob/main/PeriodogramAnalysis/PowerSpectrumAnalysis.py#L178-L182. I checked frequency and power at 0Hz is 0 (1e-10).

skim0119 commented:
> the code removing bias is here: https://github.com/Qixian-Wang/NEF/blob/main/PeriodogramAnalysis/PowerSpectrumAnalysis.py#L178-L182. I checked frequency and power at 0Hz is 0 (1e-10).

No, you are just manually removing 0Hz power from the result. Read about DC component in power spectrum analysis, and how to remove the bias.

skim0119 commented:
> What do you think of this one?

The plot looks wrong to me. Do you think it looks right?

Beside, I really meant make x-axis log scale on this plot:
<img width="747" alt="image" src="https://github.com/user-attachments/assets/e8994c3d-3d37-49a5-8a5e-bd853ee0c439">
Maybe lets make subplots, top and bottom, and make one to be semilogx and the other one to be regular.

> I googled Spectrogram but seldom use log space though.

I see people using log scale.

Qixian-Wang commented:
![Chunk0_Welch_periodogram_of_channel_3](https://github.com/user-attachments/assets/31fb66ff-17db-4ae0-8b9d-ff4a66f567dc)
![Screenshot2024_10_29_093021](https://github.com/user-attachments/assets/2da12a9f-7229-45cc-83e7-6cf52c5a464a)

I have added subplots, I think the plot should be fine, in my data, the peak value of PSD should be t 0.5 HZ, row data looks like the second figure. For which part do you think is wrong?

Qixian-Wang commented:
https://github.com/Qixian-Wang/NEF/blob/main/PeriodogramAnalysis/PowerSpectrumAnalysis.py#L103-L104. I let signal -= mean(signal) to remove DC component now.

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -59,206 +55,141 @@ def __call__(self, signal: SignalType):
+                if show:
+                    plt.show()
+                if save_path is not None:
+                    plot_path = os.path.join(save_path, f"Chunk{chunk}_Comparison_figure_channel:{channel}.png")
```
No special character

skim0119 commented:
Did you replot the periodogram after removing the bias?
What is N3 sleep EEG data?

skim0119 commented:
> For which part do you think is wrong?

![image](https://github.com/user-attachments/assets/52b9b8f2-2c4e-467c-b772-7ce3a6b2f293)
This one looks wrong. Periodogram looks good now.

skim0119 commented on PeriodogramAnalysis/MAIN.py:
```diff
@@ -3,6 +3,7 @@
 from miv.core.pipeline import Pipeline
 from PeriodogramAnalysis import PeriodogramAnalysis
 from PowerSpectrumAnalysis import SpectrumAnalysis
+from miv.datasets.openephys_sample import load_data
```
remove?

Qixian-Wang commented:
> Did you replot the periodogram after removing the bias? What is N3 sleep EEG data?

Yes, Periodogram's bias was removed, as the figure I commented this morning. 
N3 sleep EEG data: the title is wrong, it is the raw data of our recording.

Qixian-Wang commented:
> Did you replot the periodogram after removing the bias? What is N3 sleep EEG data?

![Chunk0_Spectrogram_Channel_3](https://github.com/user-attachments/assets/7c822a5e-ce5e-45f1-856c-0a7d9d784694)
Do you mean this spectrogram? It is the one after removing bias. I didn't commit it before...

skim0119 commented:
That spectrogram look weird.

Can you replot the periodogram after removing bias?

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -34,9 +34,9 @@ class SpectrumAnalysis(OperatorMixin):
+    frequency_limit: list = field(default_factory=lambda: [0.5, 100])
     nperseg: int = 2048
-    noverlap: int = 1024
+    noverlap: int = 1536
```
Why is this number changing?

Qixian-Wang commented:
![Screenshot2024_11_08_204351](https://github.com/user-attachments/assets/3f7c723b-7c2a-4832-992c-1ffecf894755)
Current Spectrogram.

skim0119 commented on PeriodogramAnalysis/PeriodogramAnalysis.py:
```diff
@@ -49,7 +48,7 @@ def __call__(self, signal: Generator[Signal, None, None]) -> Tuple[DictType, Dic
 
         Parameters:
         -----------
-        signal : SignalType
```
This looks correct. Don't change it to be `Generator` type.

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -19,246 +22,202 @@ class SpectrumAnalysis(OperatorMixin):
         super().__init__()
 
     @cache_call
-    def __call__(self, signal: SignalType):
```
Keep the `SignalType`

Qixian-Wang replied:
Sure, have changed them.

Qixian-Wang replied:
All codes have passed mypy as well.

skim0119 commented on PeriodogramAnalysis/MAIN.py:
```diff
@@ -15,26 +18,22 @@
+        band_display = [0, 5]
     )
 
+bandpass_filter: Operator = ButterBandpass(lowcut=0.2, highcut=200, order=4, tag="bandpass")
```
You made the bandpass and never used it...Is the result including the bandpass or not?

Qixian-Wang replied:
The result didn't include it, have removed it.

skim0119 replied:
I thought you applied bandpass before passing signal into spectrum analysis, no?

Qixian-Wang replied:
No, I didn't link band pass to the pipeline. When plotting Spectrum, I used filter for a while, then I removed it, the plot doesn't change much.

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -19,246 +24,254 @@ class SpectrumAnalysis(OperatorMixin):
+
+        return psd_dict, spec_dict
+
+    def computing_multi_spectrum(self, signal: Signal) -> Dict[int, Dict[str, Any]]:
```
????

skim0119 replied:
This repeats all the spectrum analysis from other operator

Qixian-Wang replied:
Have got it removed, I moved all PSD computing methods to a separate file, in commit [bc7e61d]

skim0119 commented on PeriodogramAnalysis/PeriodogramAnalysis.py:
```diff
@@ -1,282 +1,345 @@
+from MultitaperPowerSpectrum import multitaper_psd
 
-@ dataclass()
+DictType = Dict[int, Dict[int, Dict[str, Any]]]
```
Don't do this. It is impossible to see from the name that `DictType` as nested dictionary, and can be misleading

Qixian-Wang replied:
ok, I have directly used Dict[int, Dict[int, Dict[str, Any]]], previously I thought it is too long.

skim0119 commented:
@Qixian-Wang Still each method seems to be too entangled. Code is repeated between PeriodogramAnalysis and PowerSpectrumAnalysis, and it makes the unittest impossible to understand. Remember that if unittest becomes cluttered, it probably means your implementation is messy. Please clean up as the following:

- [x] Lets make separate operator for `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, `SpectrumAnalysisMultitaper`. All should have same return signature for `__call__` function.
  - [x] `plot_spectrum_methods` can be reused for all three.
- [x] Make operator `SpectrogramAnalysis` that takes either `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, or `SpectrumAnalysisMultitaper`, and run `plot_spectrogram` and `computing_spectrum`
- [x] Make another operator `PowerSpectrumAnalysis` that run `computing_ratio_and_bandpower` and its plot.

```mermaid
graph LR
P1(signal)
P2[ButterBandpass]

A1[SpectrumAnalysisWelch]
A2[SpectrumAnalysisPeriodogram]
A3[SpectrumAnalysisMultitaper]
        
B1[[SpectrogramAnalysis Operator]]
B2[[PowerSpectrumAnalysis]]
    
P1 --> P2
P2 --> A1
P2 --> A2
P2 --> A3
    
A1 --> B1
A2 --> B1
A3 --> B1
A1 --> B2
A2 --> B2
A3 --> B2
```

skim0119 commented on PeriodogramAnalysis/PowerSpectrumAnalysis.py:
```diff
@@ -19,246 +24,254 @@ class SpectrumAnalysis(OperatorMixin):
-    noverlap: int = 1024
-    convert_db: bool = True
+
+    exclude_channel_list: list[int] = field(default_factory=list)
+    band_display: list[float] = field(default_factory=lambda: [0, 100])
+    window_length_for_welch: int = 4
+    frequency_limit: list[float] = field(default_factory=lambda: [0.5, 100])
+    plotting_interval: list[float] = field(default_factory=lambda: [0, 60])
```
Just use tuple

Qixian-Wang replied:
Have changed it.

skim0119 commented on PeriodogramAnalysis/MultitaperPowerSpectrum.py:
```diff
@@ -0,0 +1,86 @@
+from scipy.signal.windows import dpss
+
+
+def multitaper_psd(
```
Where did you get this code?? Is this tested?

skim0119 commented on PeriodogramAnalysis/MultitaperPowerSpectrum.py:
```diff
@@ -0,0 +1,86 @@
+    x, sfreq, fmin=0.0, fmax=np.inf, bandwidth=None, adaptive=True, low_bias=True
+):
+    """
+    Compute multitaper power spectral density (PSD) of a given signal.
```
Which reference are you using for multitapered power spectrum?

Qixian-Wang replied:
I added Parts of this code is inspired from MNE: https://mne.tools/stable/index.html in docstring, it used similar methods as MNE.

Qixian-Wang commented:
> 



> @Qixian-Wang Still each method seems to be too entangled. Code is repeated between PeriodogramAnalysis and PowerSpectrumAnalysis, and it makes the unittest impossible to understand. Remember that if unittest becomes cluttered, it probably means your implementation is messy. Please clean up as the following:
> 
> * [x]  Lets make separate operator for `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, `SpectrumAnalysisMultitaper`. All should have same return signature for `__call__` function.
>   
>   * [x]  `plot_spectrum_methods` can be reused for all three.
> * [x]  Make operator `SpectrogramAnalysis` that takes either `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, or `SpectrumAnalysisMultitaper`, and run `plot_spectrogram` and `computing_spectrum`
> * [x]  Make another operator `PowerSpectrumAnalysis` that run `computing_ratio_and_bandpower` and its plot.
> 
> ```mermaid
> graph LR
> P1(signal)
> P2[ButterBandpass]
> 
> A1[SpectrumAnalysisWelch]
> A2[SpectrumAnalysisPeriodogram]
> A3[SpectrumAnalysisMultitaper]
>         
> B1[[SpectrogramAnalysis Operator]]
> B2[[PowerSpectrumAnalysis]]
>     
> P1 --> P2
> P2 --> A1
> P2 --> A2
> P2 --> A3
>     
> A1 --> B1
> A2 --> B1
> A3 --> B1
> A1 --> B2
> A2 --> B2
> A3 --> B2
> ``


For this part, I have separated 'SpectrumAnalysisWelch', 'SpectrumAnalysisPeriodogram', 'SpectrumAnalysisMultitaper'. But for point2, 'plot_spectrogram' and 'computing_spectrum' only use raw data, not the data after spectrum analysis, do you mean only run 'plot_spectrogram' and 'computing_spectrum' in PeriodogramAnalysis.py?

Qixian-Wang commented:
> @Qixian-Wang Still each method seems to be too entangled. Code is repeated between PeriodogramAnalysis and PowerSpectrumAnalysis, and it makes the unittest impossible to understand. Remember that if unittest becomes cluttered, it probably means your implementation is messy. Please clean up as the following:
> 
> * [x]  Lets make separate operator for `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, `SpectrumAnalysisMultitaper`. All should have same return signature for `__call__` function.
>   
>   * [x]  `plot_spectrum_methods` can be reused for all three.
> * [ ]  Make operator `SpectrogramAnalysis` that takes either `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, or `SpectrumAnalysisMultitaper`, and run `plot_spectrogram` and `computing_spectrum`
> * [ ]  Make another operator `PowerSpectrumAnalysis` that run `computing_ratio_and_bandpower` and its plot.
> 
> ```mermaid
> graph LR
> P1(signal)
> P2[ButterBandpass]
> 
> A1[SpectrumAnalysisWelch]
> A2[SpectrumAnalysisPeriodogram]
> A3[SpectrumAnalysisMultitaper]
>         
> B1[[SpectrogramAnalysis Operator]]
> B2[[PowerSpectrumAnalysis]]
>     
> P1 --> P2
> P2 --> A1
> P2 --> A2
> P2 --> A3
>     
> A1 --> B1
> A2 --> B1
> A3 --> B1
> A1 --> B2
> A2 --> B2
> A3 --> B2


I have separated SpectrumAnalysisWelch, SpectrumAnalysisMultitaper and SpectrumAnalysisWelch to a different file. In PowerSpectrumAnalysis class, only do computing_ratio_and_bandpower and plot periodogram, PSD dictionary from either methods can be the input. And in SpectrogramAnalysis class, only use raw data to plot the spectrogram. I have removed entangled periodogram calculations.

> P1 --> P2
> P2 --> A1
> P2 --> A2
> P2 --> A3
>     
> P1 --> B1
> A1 --> B2
> A2 --> B2
> A3 --> B2

Qixian-Wang commented:
> @Qixian-Wang Still each method seems to be too entangled. Code is repeated between PeriodogramAnalysis and PowerSpectrumAnalysis, and it makes the unittest impossible to understand. Remember that if unittest becomes cluttered, it probably means your implementation is messy. Please clean up as the following:
> 
> * [x]  Lets make separate operator for `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, `SpectrumAnalysisMultitaper`. All should have same return signature for `__call__` function.
>   
>   * [x]  `plot_spectrum_methods` can be reused for all three.
> * [ ]  Make operator `SpectrogramAnalysis` that takes either `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, or `SpectrumAnalysisMultitaper`, and run `plot_spectrogram` and `computing_spectrum`
> * [ ]  Make another operator `PowerSpectrumAnalysis` that run `computing_ratio_and_bandpower` and its plot.
> 
> ```mermaid
> graph LR
> P1(signal)
> P2[ButterBandpass]
> 
> A1[SpectrumAnalysisWelch]
> A2[SpectrumAnalysisPeriodogram]
> A3[SpectrumAnalysisMultitaper]
>         
> B1[[SpectrogramAnalysis Operator]]
> B2[[PowerSpectrumAnalysis]]
>     
> P1 --> P2
> P2 --> A1
> P2 --> A2
> P2 --> A3
>     
> A1 --> B1
> A2 --> B1
> A3 --> B1
> A1 --> B2
> A2 --> B2
> A3 --> B2


I have separated `SpectrumAnalysisWelch`, `SpectrumAnalysisMultitaper` and `SpectrumAnalysisWelch` to a different file. In `PowerSpectrumAnalysis` class, only do `computing_ratio_and_bandpower` and plot periodogram, PSD dictionary from either methods can be the input. And in `SpectrogramAnalysis class`, only use raw data to plot the spectrogram. I have removed entangled periodogram calculations.

```mermaid
graph LR
P1(signal)
P2[ButterBandpass]

A1[SpectrumAnalysisWelch]
A2[SpectrumAnalysisPeriodogram]
A3[SpectrumAnalysisMultitaper]
        
B1[[SpectrogramAnalysis Operator]]
B2[[PowerSpectrumAnalysis]]
    
P1 --> P2
P2 --> A1
P2 --> A2
P2 --> A3
    
P1 --> B1
A1 --> B2
A2 --> B2
A3 --> B2
```

This change is in [0484ae2] and [681839f]

skim0119 commented on PeriodogramAnalysis/SpectrogramAnalysis.py:
```diff
@@ -0,0 +1,161 @@
+        Number of points to overlap between segments for spectrogram.
+    """
+
+    frequency_limit: Tuple[float, float] = field(default_factory=lambda: [0.5, 100])
```
Use tuple as default

Qixian-Wang replied:
solved them in commit [04c83fd]

Qixian-Wang commented:
Changed `Power_Spectral_Density.py` in commit [a80a7b6], it is more decent now.

skim0119 commented:
@Qixian-Wang I'll check the code. Can you re-write the PR? Use diagram above, and give more detail description on what features are added. Also check the github-action.

skim0119 commented on PeriodogramAnalysis/main.py:
```diff
@@ -0,0 +1,39 @@
+from miv.core.operator import DataLoader
```
Remove unused modules

Qixian-Wang replied:
Got useless modules removed in all files.

skim0119 commented on PeriodogramAnalysis/main.py:
```diff
@@ -0,0 +1,39 @@
+from miv.core.operator import Operator, DataLoader
+from power_density_statistics import SpectrumAnalysisPeriodogram, SpectrumAnalysisMultitaper, SpectrumAnalysisWelch
+
+working_directory = "results"
```
Use the sample dataset instead of your own.

Qixian-Wang replied:
Have changed it in [aeb5343]

skim0119 commented on PeriodogramAnalysis/multitaper_spectrum_statistics.py:
```diff
@@ -0,0 +1,87 @@
+from scipy.signal.windows import dpss
+
+
+def multitaper_psd(
```
Make sure you test this function, especially cuz you said it is generated code.

Qixian-Wang replied:
Have tested it, the algorithm is fine and I replaced this function with mne function, the plots are the same.

Qixian-Wang replied:
my bad, previously the value of mne is twice the value of our function, it should because `rfft` returned psd of only positive frequency, I changed it in commit [276e6aa]

skim0119 replied:
I don't see the test code

Qixian-Wang replied:
Added relevant tests in this commit: [1e2470a]

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+
+
+@dataclass
+class SpectrumAnalysisBase(OperatorMixin):
```
Add documentation

Qixian-Wang replied:
Have added.

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+@dataclass
+class SpectrumAnalysisWelch(SpectrumAnalysisBase):
+    window_length_for_welch: int = 4
+    band_display: Tuple[float, float] = (0, 100)
```
defined above

Qixian-Wang replied:
Document added.

skim0119 replied:
> "defined" above

Qixian-Wang replied:
Ah, get it, have removed
 `window_length_for_welch: int = 4
    band_display: Tuple[float, float] = (0, 100)`
to base class

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+
+@dataclass
+class SpectrumAnalysisPeriodogram(SpectrumAnalysisBase):
+    window_length_for_welch: int = 4
```
Not used

Qixian-Wang replied:
Got deleted

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+        psd_periodogram_dict: Dict[int, Dict[str, Any]] = defaultdict(dict)
+
+        for channel in range(signal.number_of_channels):
+            signal_no_bias = signal.data[:, channel] - np.mean(signal.data[:, channel])
```
Like I mentioned before, don't access `signal.data` directly. It is because the data axis is not consistent among different recording devices. It will be taken care internally, you just use `[]` operator to access data for each channel.

Use `signal[channel]`. You can probably just do
```py
for channel_signal in signal:
```

Qixian-Wang replied:
removed them in commit [f46e5a5]

Qixian-Wang replied:
Is this fine to use:     `for channel, channel_signal in enumerate(signal):` ?

skim0119 replied:
> Is this fine to use: `for channel, channel_signal in enumerate(signal):` ?

no. `enumerate` just returns index, they are not channel

Qixian-Wang replied:
but I don't have exclude channel list, how can I know which channels are removed?

skim0119 replied:
No channels are removed.

Qixian-Wang replied:
I have changed it in commit [322ed94]

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+    @cache_call
+    def __call__(self, signal: SignalType) -> Dict[int, Dict[int, Dict[str, Any]]]:
+
+        psd_dict: Dict[int, Dict[int, Dict[str, Any]]] = defaultdict(dict)
```
Do you have to use a nested dictionary for this?? Isn't it just 2D array?

Qixian-Wang replied:
The dict structure is [chunk][channel]{"freqs", "psd"}, it should be necessary.

skim0119 replied:
But chunk and channels are just indices. Channels could be non incremental, but chunk is literally just index.

Qixian-Wang replied:
I have simplified the dicts, they are now in the type: `Dict[int, Dict[str, Any]]`

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+        psd_dict = output
+        for chunk in psd_dict.keys():
+            for channel in psd_dict[chunk].keys():
+                plt.stem(
```
We need this?? Isn't it just same as `plot` with all this `fmt`s?

Qixian-Wang replied:
Have removed it, it is just copied from the original paper, their sampling rate is much lower than ours, so it will the figure better in their case.

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+                if save_path is not None:
+                    plot_path = os.path.join(
+                        save_path,
+                        f"Chunk{chunk}_{self.tag.replace(' ', '_').lower()}_channel_{channel}.png",
```
Dont add `self.tag` yourself. it is included as part of `save_path`.

Also, make different folder for each channel, instead of adding it to a filename. It will create too many images in single folder.

```
save_path/
│
├── channel01/
│   ├── power_density_01.png
│   ├── power_density_02.png
│   └── power_density_03.png
│
├── channel02/
    ├── power_density_01.png
    ├── power_density_02.png
    └── power_density_03.png
```

Qixian-Wang replied:
Have changed all the paths in previous commits.

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+                        f"Chunk{chunk}_{self.tag.replace(' ', '_').lower()}_channel_{channel}.png",
+                    )
+                    plt.savefig(plot_path, dpi=300)
+                plt.close()
```
`plt.close('all')`

Qixian-Wang replied:
got them changed.

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+from scipy import signal as sig
+from dataclasses import dataclass, field
+from collections import defaultdict
+from typing import Tuple, Dict, Any, Generator, Optional
```
Lets keep `from typing import ...` on the first line. It is common convension

skim0119 replied:
Also I'm not sure you know, `Tuple`, `Dict`, `List` are deprecated. You can just directly use `list`, `tuple`, etc.

Qixian-Wang replied:
Ah I see, I have moved `from typing import ...` to the first line, and replaced `Tuple`, etc.

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+class SpectrumAnalysisWelch(SpectrumAnalysisBase):
+    window_length_for_welch: int = 4
+    band_display: Tuple[float, float] = (0, 100)
+    tag: str = "Welch PSD"
```
`Welch PSD spectrum analysis

Qixian-Wang replied:
Have changed all tags in this file.

skim0119 commented on PeriodogramAnalysis/power_density_statistics.py:
```diff
@@ -0,0 +1,125 @@
+        return psd_dict
+
+    def compute_psd(self, signal: Signal) -> Dict[int, Dict[str, Any]]:
+        raise NotImplementedError("compute_psd not finished")
```
Change msg to indecate that this `Base` class is not meant to be used standalone, and user should use `SpectrumAnalysisWelch`, `SpectrumAnalysisPeriodogram`, etc..

Qixian-Wang replied:
Got it changed.

skim0119 commented on PeriodogramAnalysis/spectrogram_analysis.py:
```diff
@@ -0,0 +1,161 @@
+
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy import signal as sig
```
Please change `sig`. I think it is quiet confusing. You have so many `signal` inside the code, it is not obvious that `sig` is module from `scipy`. I would either just use full `scipy.signal` or name it something doesn't remind other stuffs, like `sps` (scipy-signal).

Qixian-Wang replied:
Have changed them to `scipy.signal.welch`, etc. Forgot it can be used like this....

skim0119 commented on PeriodogramAnalysis/spectrogram_analysis.py:
```diff
@@ -0,0 +1,161 @@
+                    plt.show()
+                if save_path is not None:
+                    plot_path = os.path.join(
+                        save_path, f"Chunk{chunk}_Spectrogram_Channel_{channel}.png"
```
same as above. Lets keep different folder for each channel, and `chunk` can be added as index. `save_path/spectrogram_{chunk:03d}.png`

Qixian-Wang replied:
Sure, have changed the path and name.

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+
+    Attributes:
+    -----------
+    band_list : list
```
```suggestion
    band_list : tuple[tuple[int, int], ...]
```

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+    Attributes:
+    -----------
+    band_list : list
+        Frequency bands to analyze, default is [[0.5, 4], [4, 8], [8, 12], [12, 30], [30, 100]], i.e. the five common frequency bands in EEG.
```
Don't need to mention EEG

```suggestion
        Frequency bands to analyze.
        (default: ((0.5, 4), (4, 8), (8, 12), (12, 30), (30, 100)) )
```

Qixian-Wang replied:
Removed

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+        Frequency bands to analyze, default is [[0.5, 4], [4, 8], [8, 12], [12, 30], [30, 100]], i.e. the five common frequency bands in EEG.
+    """
+
+    band_list: tuple = ((0.5, 4), (4, 8), (8, 12), (12, 30), (30, 100))
```
```suggestion
    band_list: tuple[tuple[int, int], ...] = ((0.5, 4), (4, 8), (8, 12), (12, 30), (30, 100))
```

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+        power_dict
+            power dictionary
+        """
+        power_dict: Dict[int, Dict[int, Dict[str, Any]]] = defaultdict(dict)
```
Again, not sure if nested dictionary is a good data structure here.

Qixian-Wang replied:
I hope power_dict can contain data from all chunks, so I think this nested dict is needed. And psd_dict returned by power_density_statistics.py contains psd data of all chunks.

skim0119 replied:
Lets discuss this above

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+        power_dict
+            Dictionary containing power values for all channels in this chunk.
+        """
+        bands = {
```
Why not use `self.band_list`?

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+        }
+        power_dict: Dict[int, Dict[str, Any]] = defaultdict(dict)
+
+        for channel in psd_dict.keys():
```
Clean this code. User will define the `band_list`.

You don't need to worry about the name behind the band (i.e. delta, theta, etc.). Those name don't mean much for in-vitro electrophysiology.

Qixian-Wang replied:
I see, I use self.band_list in all cases now, in commit [aeb5343]

skim0119 commented on PeriodogramAnalysis/spectrum_analysis.py:
```diff
@@ -0,0 +1,318 @@
+                if save_path is not None:
+                    plot_path = os.path.join(
+                        save_path,
+                        f"Chunk{chunk}_periodogram_of_channel_{channel}.png",
```
Again, rename

Qixian-Wang replied:
Have changed them in commit [aeb5343]

skim0119 commented on PeriodogramAnalysis/test_multitaper_spectrum_statistics.py:
```diff
@@ -0,0 +1,31 @@
+from miv.io.openephys import Data, DataManager
+from multitaper_spectrum_statistics import multitaper_psd
+
+path: str = load_data(progbar_disable=True).data_collection_path
```
Don't define variable outside the scope, especially for testing

Qixian-Wang replied:
I see, moved it in this commit [f5695ac]

skim0119 commented on PeriodogramAnalysis/test_multitaper_spectrum_statistics.py:
```diff
@@ -0,0 +1,31 @@
+            sum_pds = np.sum(np.abs(psd_mne))
+
+            # Test the difference is smaller than 1%
+            assert sum_diff_psd / sum_pds < 1e-2
```
`1e-2`? I thought your implementation is almost same as the one in `mne`, wouldn't it needs to be the same or almost the same?

Qixian-Wang replied:
I have modified it, I copied some functions from `mne` and now they are almost the same. I used `np.testing.assert_allclose(psd, psd_mne, rtol=1e-3)` and it passes, when set rtol to 1e-4, only 5 points in 150000 will not pass.

skim0119 commented on PeriodogramAnalysis/test_multitaper_spectrum_statistics.py:
```diff
@@ -0,0 +1,31 @@
+            )
+
+            difference_psd = psd_mne - psd
+            sum_diff_psd = np.sum(np.abs(difference_psd))
```
Why average and compare? You are adding more noise in the process. Just use `np.testing.assert_allclose` or something equivalent.

Qixian-Wang replied:
Have changed, check commit [d7a26a7]

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,80 @@
+    yield signal2
+
+
+def test_SpectrumAnalysisWelch(tmp_path):
```
I don't get these tests. You created mock signals, but all assertion you are making in the test is whether `psd_dict` is dictionary and if `freqs` and `psd` is in the result??

Qixian-Wang replied:
Yeah, it is weird... have added numerical tests in these commits: [d7a26a7] [424b207]

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,134 @@
+
+                idx = np.argmin(np.abs(freqs - exp_freq))
+                # Test the peaks of psd is as expected
+                assert freqs[idx] == exp_freq or np.isclose(
```
Can you justify atol? You know, atol=0.1 basically means number can be quiet different

skim0119 replied:
I would suggest just use default: atol and rtol

Qixian-Wang replied:
Sure, I changed to default, and it passed the tests.

skim0119 commented on PeriodogramAnalysis/test_spectrum_analysis.py:
```diff
@@ -0,0 +1,147 @@
+    Test PowerSpectrumAnalysis class to ensure it computes the power_dict correctly
+    based on the provided psd_dict, and returns the expected values.
+    """
+    psd_dict = mock_psd_list()
```
just call it psd_input or something. It is not mocking anything

Qixian-Wang replied:
Sure, changed to psd_input, also modified similar names in other tests.

skim0119 commented on PeriodogramAnalysis/test_spectrum_analysis.py:
```diff
@@ -0,0 +1,147 @@
+        computed_power = result_power_dict["power_list"][i]
+        computed_rel_power = result_power_dict["rel_power_list"][i]
+
+        assert np.isclose(manual_power, computed_power, rtol=1e-2)
```
rtol is to big. why?

Qixian-Wang replied:
I previously remembered default is like 1e-1, I have changed `rtol` and `atol` to default, still passed tests.

Qixian-Wang replied:
Also replaced others with default, I also modified multitaper method, now it produce exact the same results as `mne`.

Qixian-Wang replied:
Updated these changes here: [d124711]

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,237 @@
+    )
+
+
+def test_SpectrumAnalysisBase_call(mocker) -> None:
```
Separate out the test.

Test to check parameters: For this, you need to test negatives: integer window length, non-negative band, ordered band (band[0] < band[1]) etc. If your function runs with wrong parameters, your code is not safe. Here, you are just testing if the default value is the same, which doesn't give any useful test for two reasons: your test depends on your implementation which means your test needs to be adapted to your implementation. It is not how the test should be written. Secondly, if this test fails, it just means your test is wrong, not your implementation. Think about the purpose of writing unittest.

Qixian-Wang replied:
I have added more negative tests for parameters: https://github.com/Qixian-Wang/NEF/blob/main/PeriodogramAnalysis/spectrogram_analysis.py#L38-L55.     I am not sure if it is too much though.

Qixian-Wang replied:
Revision to this test_file is in this commit: [8dcc53a]
Revisions to the rest is here: [1a0b4a9]

Also, I moved plotting function in `power_density_statistics.py` to `spectrum_analysis.py`. One reason is previously when I use generator module, the generator will be consumed during plotting in `power_density_statistics.py`, so no data will be collected in following modules. Second is I think the figures plotted by  `power_density_statistics.py` is also periodogram. I am also wondering if we should remove it.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,237 @@
+    to test if the dict returned by call is correct, if it fails, the call function may be incorrect
+    the functions may not be properly called and the dict's layers might be wrong.
+
+    I mock "compute_psd" and see how many times it is called to see if it is used to handle all chunks.
```
Don't mock `compute_psd`, make mock class that inherit `SpectrumAnalysisBase`. Isn't it how user suppose to use?

Qixian-Wang replied:
I see, have deleted it, it is not necessary right now.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,237 @@
+
+    psd_dict_list = [psd_dict_wel, psd_dict_per]
+
+    for psd_dict in psd_dict_list:
```
I still disagree you using dictionary for representing a list of chunks. Dictionary is unordered data structure, where chunk intrinsically means ordered structure. Outputing dictionary simply doesn't make sense, and it is error-prone: you need to make sure that downstream receiver receives the data in order.

If you want to process each chunk at a time, I would make generator module instead.

Qixian-Wang replied:
I have removed all dicts and used generator module in this commit [6a99f19], for some part ,I used lists to save data for plotting.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def test_SpectrumAnalysisPeriodogram() -> None:
+    """
```
If it is exactly same test as above, try not to repeat the code. Use `pytest.mark.parametrize`.

I like the test method. Maybe include negative test: low power for non-desire frequency.

Qixian-Wang replied:
Have implemented pytest.mark.parametrize and added negative test.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,237 @@
+            assert range_psd_sum / total_psd_sum >= 0.9
+
+
+def test_SpectrumAnalysisWelch_compute_psd() -> None:
```
This test would be not needed if you use handle the signal processing within the generator.

Qixian-Wang replied:
Have deleted it now.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,237 @@
+    assert len(result) == 2
+
+
+def test_SpectrumAnalysisPeriodogram_compute_psd() -> None:
```
Don't copy and paste code.

If you see same code, it means it can be abstracted.

Qixian-Wang replied:
Have used pytest.mark.parametrize to get rid of duplication.

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -0,0 +1,84 @@
+from spectrogram_analysis import SpectrogramAnalysis
+
+
+def signal_input() -> Generator:
```
Again, don't copy and paste code.

Qixian-Wang replied:
Have removed all similar codes.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+)
+
+
+def mock_signal_generator():
```
`pytest.fixture`

Qixian-Wang replied:
Have added `pytest.fixture` to all three `test_` code.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+    assert len(result) == 2
+    assert isinstance(freqs, np.ndarray)
+    assert isinstance(psd, np.ndarray)
+    assert freqs.shape[0] == 801
```
Where is this number coming from??

Qixian-Wang replied:
It is calculated based on window length, sampling rate and FFT nodes.

Δf = sampling_rate / nfft = 100 / 1600 = 0.0625 Hz
Frequency Range = [0, fs/2] = [0, 50] Hz
Number of Frequency Points = (50 - 0) / Δf + 1 = 801

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+    for channel in range(2):
+        psd_channel = psd[:, channel]
+        total_psd_sum = np.sum(psd_channel)
+        expected_freqs = [5, 10, 15, 20]
```
This should be returned along `mock_signal_generator` so that if you want to add additional test signal later, you only need to modify one time.

Qixian-Wang replied:
It make sense, I have moved `expected_freqs` to `mock_signal_generator`

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+        + np.sin(2 * np.pi * 20 * timestamps)
+    )
+
+    data1 = np.array([signal, signal]).T
```
```suggestion
    data1 = np.array([signal, 2*signal]).T
```

Qixian-Wang replied:
Have changed it.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+        # Test the psd around peaks comprises more than 90% of total psd
+        assert range_psd_sum / total_psd_sum >= 0.9
+
+        for i, freq in enumerate(freqs):
```
Double check this. It looks weird

Qixian-Wang replied:
I checked the data, should be fine.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+            # Test the peaks of psd is as expected
+            assert np.isclose(freqs[idx], exp_freq)
+        # Test the psd around peaks comprises more than 90% of total psd
+        assert range_psd_sum / total_psd_sum >= 0.9
```
would this be true if you have noisy data?

Qixian-Wang replied:
I added some gaussian noise to `mock_signal_generator`, it still passes the test. I checked psd data and they have very high concentration.

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -0,0 +1,96 @@
+    [SpectrumAnalysisWelch, SpectrumAnalysisPeriodogram],
+)
+def test_different_sampling_rates(analysis_class):
+    for rate in [50, 200, 500]:
```
This test looks clean. I would use pytest.mark.parametrize for `rate` as well

Qixian-Wang replied:
Sure, have used `pytest.mark.parametrize` for `rate`.

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -0,0 +1,73 @@
+from spectrogram_analysis import SpectrogramAnalysis
+
+
+def signal_input():
```
I see some copy and paste here

Qixian-Wang replied:
Forgot `mock_signal_generator` can be imported.. now this signal generator is imported here `from test_power_density_statistics import mock_signal_generator`

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -0,0 +1,73 @@
+        channel_data = signal_gen.data[:, ch]
+
+        signal_no_bias = channel_data - np.mean(channel_data)
+        freqs_expected, times_expected, sxx_expected = spectrogram(
```
Aren't you just repeating the code here?
If you think the `spectrogram` from `scipy` is working properly, just check if this function is called with correct parameters, using `mock`, and results match the expected dimensions. No need to re-run `spectrogram`, you are not testing anything as it is.

Qixian-Wang replied:
Have changed it, previously I was thinking it is just repeating as well but not sure how to fix it.

skim0119 commented on PeriodogramAnalysis/test_spectrum_analysis.py:
```diff
@@ -0,0 +1,93 @@
+        freqs, psd
+    )
+
+    for ch in range(2):
```
I see you have two channels, but I don't see why. Just use 1 channel and check separately if multi-channel result can be handled properly.

Qixian-Wang replied:
Ok, I thought testing multi-channels directly will be more efficient.

Qixian-Wang replied:
I changed the code here [9aaf6f8] but not sure if it is reasonable.

skim0119 commented on PeriodogramAnalysis/test_spectrum_analysis.py:
```diff
@@ -0,0 +1,93 @@
+from scipy.integrate import simpson
+
+
+def psd_input() -> dict[int, dict[str, Any]]:
```
consider using `fixture`

Qixian-Wang replied:
Have implemented fixture

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -75,22 +81,28 @@ def test_spectrum_analysis_computation(analysis_class) -> None:
 @pytest.mark.parametrize(
-    "analysis_class",
-    [SpectrumAnalysisWelch, SpectrumAnalysisPeriodogram],
+    "analysis_class, rate",
```
fix this
<img width="825" alt="image" src="https://github.com/user-attachments/assets/eeaad328-835a-4eed-b77b-717368004a24" />

Qixian-Wang replied:
Have fixed it.

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -5,20 +5,7 @@
-    data1 = np.array([signal, signal]).T
-    signal = Signal(data=data1, timestamps=timestamps, rate=100)
-    return signal
+from test_power_density_statistics import mock_signal_generator
```
https://docs.pytest.org/en/stable/how-to/fixtures.html#scope-sharing-fixtures-across-classes-modules-packages-or-session

Qixian-Wang replied:
Good to know, I have implemented it in commit [9e1fd9b]

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -52,22 +40,36 @@ def test_call_return_shape():
+    mock_spectrogram = mocker.patch("scipy.signal.spectrogram")
+
+    mock_spectrogram.return_value = (
+        np.random.rand(13),
```
what is this? Avoid using random in any unittest. Also what is 13 and 76?

Qixian-Wang replied:
Added explanation of 13 and 76, and I used np.ones now, it should still be representative. They are in commit [9e1fd9b]

skim0119 commented on PeriodogramAnalysis/test_power_density_statistics.py:
```diff
@@ -18,16 +20,19 @@ def mock_signal_generator():
     )
-
-    data1 = np.array([signal, signal]).T
+    noise = np.random.normal(loc=0, scale=0.5, size=signal.shape)
```
avoid using random. Fix seed at least

Qixian-Wang replied:
Used seed in stead.

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -0,0 +1,75 @@
+        args, kwargs = call
+
+        expected_signal = signal.data[:, ch] - np.mean(signal.data[:, ch])
+        np.testing.assert_array_equal(args[0], expected_signal)
```
This is good test.

skim0119 commented on PeriodogramAnalysis/test_spectrogram_analysis.py:
```diff
@@ -0,0 +1,75 @@
+            (int(signal.rate * power_analysis.nperseg_ratio)) / 2
+        )
+
+    # Test the return
```
This can be separate test

Qixian-Wang replied:
Have separated them in commit [9e1fd9b]
